components/modules/rmt.c

```diff
@@ -0,0 +1,403 @@
+  int gpio = luaL_checkinteger(L, 1);
+  int divisor = get_divisor(L, 2);
+
+  // We will set the channel later
+  rmt_config_t config = RMT_DEFAULT_CONFIG_RX(gpio, 0);
+  config.clk_div = divisor;
+  config.rx_config.idle_threshold = 65535;
+
+  if (lua_type(L, 3) == LUA_TTABLE) {
+    lua_getfield(L, 3, "invert");
+    if (lua_toboolean(L, -1)) {
+      config.flags |= RMT_CHANNEL_FLAGS_INVERT_SIG;
+    }
+    lua_pop(L, 1);
+
+    lua_getfield(L, 3, "filter_ticks");
+    if (!lua_isnil(L, -1)) {
+      int ticks = lua_tointeger(L, -1);
+      if (ticks < 0 || ticks > 255) {
+        return luaL_error(L, "filter_ticks must be in the range 0 - 255");
+      }
+      config.rx_config.filter_ticks_thresh = ticks;
+      config.rx_config.filter_en = true;
+    }
+    lua_pop(L, 1);
+
+    lua_getfield(L, 3, "idle_threshold");
+    if (!lua_isnil(L, -1)) {
+      int threshold = lua_tointeger(L, -1);
+      if (threshold < 0 || threshold > 65535) {
+        return luaL_error(L, "idle_threshold must be in the range 0 - 65535");
+      }
+      config.rx_config.idle_threshold = threshold;
+    }
+    lua_pop(L, 1);
+  }
+
+  lrmt_channel_t ud = (lrmt_channel_t)lua_newuserdata(L, sizeof(*ud));
+  if (!ud) return luaL_error(L, "not enough memory");
+  memset(ud, 0, sizeof(*ud));
+  luaL_getmetatable(L, "rmt.channel");
+  lua_setmetatable(L, -2);
+
+  // We have allocated the channel -- must free it if the rest of this method fails
+  int channel = platform_rmt_allocate(1, RMT_MODE_RX);
+
+  if (channel < 0) {
+    return luaL_error(L, "no spare RMT channel");
+  }
+
+  config.channel = channel;
+
+  ud->channel = channel;
+  ud->tx = false;
+
+  int rc = rmt_config(&config);
+  if (rc) {
+    platform_rmt_release(config.channel);
+    return luaL_error(L, "Failed to configure RMT");
+  }
+
+  rc = rmt_driver_install(config.channel, 1000, 0);
+  if (rc) {
+    platform_rmt_release(config.channel);
+    return luaL_error(L, "Failed to install RMT driver");
+  }
+
+  lua_pushinteger(L, divisor * 12500);
+
+  return 2;
+}
+
+static void free_transmit_wait_params(lua_State *L, lrmt_cb_params *p) {
+  if (!p->data) {
+    luaL_unref(L, LUA_REGISTRYINDEX, p->cb_ref);
+    luaL_unref(L, LUA_REGISTRYINDEX, p->data_ref);
+  }
+  free(p);
+}
+
+static void handle_receive(void *param) {
+  lrmt_cb_params *p = (lrmt_cb_params *) param;
+
+  RingbufHandle_t rb = NULL;
+
+  //get RMT RX ringbuffer
+  rmt_get_ringbuf_handle(p->channel, &rb);
+  // Start receive
+  rmt_rx_start(p->channel, true);
+  while (!p->rx_shutting_down) {
+    size_t length = 0;
+    rmt_item32_t *items = NULL;
+
+    items = (rmt_item32_t *) xRingbufferReceive(rb, &length, 50 / portTICK_PERIOD_MS);
+    if (items && length) {
+      lrmt_cb_params *rx_params = malloc(sizeof(lrmt_cb_params) + length);
+      if (rx_params) {
+        memset(rx_params, 0, sizeof(*rx_params));
+	memcpy(rx_params + 1, items, length);
+        rx_params->cb_ref = p->cb_ref;
+	rx_params->data = (void *) items;
```

Comment:
I think you want `= (rmt_item32_t *)(rx_params + 1);` here, seeing as you've already copied out `items` and will explicitly return `items` to the ringbuffer below in `vRingbufferReturnItem()`.